(GH-1918) Add 'Debugging tasks' section to 'Writing tasks' (#2034)
puppetlabs/bolt: documentation/writing_tasks.md

For more information on securely passing user input, see the blog post [Stop
using backtick to run shell command in
Ruby](https://www.hilman.io/blog/2016/01/stop-using-backtick-to-run-shell-command-in-ruby/).

## Debugging tasks

> Because tasks are executed remotely, it can be challenging to debug tasks that

Reviewer: ```suggestion
When tasks are executed remotely, it can be challenging to debug tasks that
```

Reviewer: Or just remove the line :shrug:

are not working as expected. There are several ways that you can debug tasks,
including using remote debugging libraries, using methods available in the
task helper libraries, running the task locally as a script, and redirecting
`stderr` to `stdout`.

### Debug logs

> Typically, Bolt will only display task output that is sent to `stdout`.

Reviewer: It might be useful to add that stderr will only be printed if nothing is sent to stdout. I get that it seems like TMI, but it's come up a few times just this week in channels and seems useful to document.

Author: I think in this specific case (using `--log-level debug`) it's clearer to set an expectation that task output will usually be `stdout` and that you can always see what was sent to `stderr` in debug mode.

However, Bolt does log additional information about a task run, including
output sent to `stderr`, at the `debug` level. You can view these logs
during a task run using the `--log-level debug` CLI option.

```shell
$ bolt task run mytask param1=foo param2=bar -t all --log-level debug
```

> ### Debuggers

Reviewer 2: This section is a little confusing. It's written as if you're debugging against a remote target, but the Ruby and PS instructions only work for localhost, while the Python instructions do work for a remote target. There's just a "tip" for Ruby/PS that says it's _simpler_ to use localhost, but they don't specify that the instructions are only for localhost.

Author: Ya, this one in particular is tricky since connecting to the sockets that are opened by the Ruby/Python debuggers can be difficult. I think the benefit of running the task on localhost is that you don't need to make changes to the task to run it as a script -- I can make that clearer / less confusing in this section.

For PowerShell, James' suggestion was to always debug it as a script, since remote debugging with PowerShell can be really complicated. I can move this section under "Running a task as a script", but it seems more appropriate to have it under "Debuggers".

Many of the scripting languages you can use to write tasks have debugging

> libraries available that allow you to step through and examine your task

Reviewer: ```suggestion
libraries available that allow you to set breakpoints and examine your task
```

Reviewer: I'm just assuming this is also how the Python library works.

> `localhost` instead of a list of remote targets. This avoids the need to
> establish connections with the socket opened by `pry-remote`, which may be
> difficult if your target restricts incoming connections.

You can quickly install the `pry-remote` gem using Bolt's Ruby:

```shell
$ /opt/puppetlabs/bolt/bin/gem install --user-install pry-remote
```

If you are running Bolt on Windows with the default install location, run
the following command instead:

```powershell
> "C:/Program Files/Puppet Labs/Bolt/bin/gem.bat" install --user-install pry-remote
```

Then, add the following line to your task wherever you want to begin debugging:

```ruby
require 'pry-remote'; binding.remote_pry
```

You can then open a connection to a target using the `pry-remote` command:

```shell
$ pry-remote

Frame number: 0/4

From: /tmp/4f9dcfa3-ce0c-49e2-bcf5-8d761b202186/ruby/tasks/init.rb @ line 9 MyClass#task:

     6: def task(opts)
 =>  9:   require 'pry-remote'; binding.remote_pry
    10:
```

📖 **Related information**
- [`pry-remote` documentation](https://www.rubydoc.info/gems/pry-remote)
- [`pry` documentation](http://pry.github.io/)

> #### PowerShell tasks

Reviewer 3: ❤️

the breakpoint.

```powershell
> ./mytask.ps1
```

📖 **Related information**
- [`Set-PSBreakpoint`
  documentation](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/set-psbreakpoint?view=powershell-7)

### Task helper methods

Both the [Python task
helper](https://github.com/puppetlabs/puppetlabs-python_task_helper) and [Ruby
task helper](https://github.com/puppetlabs/puppetlabs-ruby_task_helper) include
methods that can help you debug a task. The `debug` method logs arbitrary values
as debugging messages, while the `debug_statements` method returns an array of
the logged debugging messages.

When a task using a task helper library raises a `TaskError`, the error will
include any logged debugging messages under the `details` key. You can also add
debugging statements when raising a `TaskError` yourself by calling
`debug_statements` and adding the result under the `details` key.

> 🔩 **Tip:** When running a task with `bolt task run`, use the `--format json`
> option to see the full result from the task, including the value of the
> `details` key.

#### Python tasks

The following Python task includes a few debugging statements which describe

> what the task is doing and the results from a couple arithmetic operations:

Reviewer: We should also include the `task.json` for this so that the helper files load correctly.

```python
#!/usr/bin/env python

import os, sys
sys.path.append(os.path.join(os.path.dirname(__file__), '..', '..', 'python_task_helper', 'files'))
from task_helper import TaskHelper

class MyTask(TaskHelper):
    def task(self, args):

>       self.debug('Adding values')

Reviewer: ```suggestion
        self.debug('Adding values')
```
Python typically has 4 space tabs

Author: Remind me again why so many people love Python?

    debug "Quotient of values: #{quotient}"

    { sum: sum, quotient: quotient }
  end
end

if __FILE__ == $0
  MyTask.run
end
```

Running this task with the parameter `value_2=0` will raise a `TaskError`
that will automatically include the logged debugging statements under the
`details` key:

```shell
$ bolt task run mytask -t localhost value_1=10 value_2=0 --format json

{
  "items":[
    {
      "target":"localhost",
      "action":"task",
      "object":"mytask",
      "status":"failure",
      "value":{
        "_error":{
          "kind":"ZeroDivisionError",
          "msg":"divided by 0",
          "details":{
            "debug":[
              "Adding values",
              "Sum of values: 1",
              "Dividing values"
            ]
          }
        }
      }
    }
  ],
  "target_count":1,
  "elapsed_time":0
}
```       

📖 **Related information**
- [Ruby task helper
  debugging](https://github.com/puppetlabs/puppetlabs-ruby_task_helper#debugging)

### Running a task as a script

Running tasks without Bolt can make debugging easier, as the task will no
longer be executed in a separate thread. Since tasks are similar to scripts,

> you can make temporary changes to the task so they can accept command line

Reviewer 4: ```suggestion
you can make temporary changes to the task so they can accept command-line
```
This is a style thing - "command-line" for adjectives, "command line" for nouns. Our docs aren't super consistent on this. I'll open a pothole PR for this at some point.

arguments and be run from the command line.

#### Python examples

Tasks that are written to accept input from `stdin` can already by run as
a script by piping the parameters as JSON to the task:

```python
#!/usr/bin/env python
import sys, json

params = json.load(sys.stdin)

result = {
  "sum": params['num1'] + params['num2']
}

print(json.dumps(result))
```

```shell

> $ echo '{"num1":10,"num2":5}' | mytask.py

Reviewer: ```suggestion
$ echo '{"num1":10,"num2":5}' | python mytask.py
```

Reviewer: or `./mytask.py`

```

> Alternatively, you can write your task to parse command line arguments

Reviewer 4: ```suggestion
Alternatively, you can write your task to parse command-line arguments
```

as parameters if they are present:

```python
#!/usr/bin/env python
import sys, json

if len(sys.argv) > 0:
  params = {
    "num1": int(sys.argv[1]),
    "num2": int(sys.argv[2])
  }
else:
  params = json.load(sys.stdin)

result = {
  "sum": params['num1'] + params['num2']
}

print(json.dumps(result))
```

```shell

> $ mytask.py 10 5

Reviewer: ```suggestion
$ python mytask.py 10 5
```

```

#### Ruby examples

Tasks that are written to accept input from `stdin` can already be run as
a script by piping the parameters as JSON to the task:

```ruby
#!/usr/bin/env ruby
require 'json'

params = JSON.parse(STDIN.read)

result = {
  'sum' => params['num1'] + params['num2']
}

puts result.to_json
```

```shell

> $ echo '{"num1":10,"num2":5}' | mytask.rb

Reviewer: ```suggestion
$ echo '{"num1":10,"num2":5}' | ruby mytask.rb
```

```

> Alternatively, you can write your task to parse command line arguments

Reviewer 4: ```suggestion
Alternatively, you can write your task to parse command-line arguments
```

as parameters if they are present:

```ruby
#!/usr/bin/env ruby
require 'json'

if ARGV.any?
  params = {
    'num1' => ARGV[0].to_i,
    'num2' => ARGV[1].to_i
  }
else
  params = JSON.parse(STDIN.read)
end

result = {
  'sum' => params['num1'] + params['num2']
}

puts result.to_json
```

```shell

> $ mytask.rb 10 5

Reviewer: ```suggestion
$ ruby mytask.rb 10 5
```

```

#### PowerShell example

Bolt sends parameters to PowerShell tasks by converting the parameters into
named arguments. You can run a PowerShell task as a script by running it
from the command line and providing the parameters as named arguments:

```powershell
[CmdletBinding()]
Param(
	[Int]$num1,
  [Int]$num2
)

$result = @{Sum=($num1 + $num2)} | ConvertTo-Json

Write-Output $result
```

```powershell
> ./mytask.ps1 -num1 10 -num2 5
```

### Redirecting `stderr`

> By default, Bolt will not display output from `stderr` if any output is sent

Reviewer: :+1:

> By default, Bolt will not display output from `stderr` if any output is sent

Reviewer 4: ```suggestion
By default, Bolt does not display output from `stderr` if any output is sent
```
